packages/playground/remote/service-worker.ts

```diff
@@ -27,8 +28,66 @@ if (!(self as any).document) {
 
 reportServiceWorkerMetrics(self);
 
+const cache = new WorkerCache(buildVersion);
+
+/**
+ * For offline mode to work we need to cache all required assets.
+ *
+ * These assets are listed in the `/assets-required-for-offline-mode.json` file
+ * and contain JavaScript, CSS, and other assets required to load the site without
+ * making any network requests.
+ */
+cache.cacheOfflineModeAssets();
+
+/**
+ * Cleanup old cache.
+ *
+ * We cache data based on `buildVersion` which is updated whenever Playground is built.
+ * So when a new version of Playground is deployed, the service worker will remove the old cache and cache the new assets.
+ *
+ * If your build version doesn't change while developing locally check `buildVersionPlugin` for more details on how it's generated.
+ */
+cache.cleanup();
+
+/**
+ * Handle fetch events and respond with cached assets if available.
+ */
+self.addEventListener('fetch', (event) => {
+	const url = new URL(event.request.url);
+
+	/**
+	 * Don't cache requests to the service worker script itself.
+	 */
+	if (url.pathname.startsWith(self.location.pathname)) {
+		return;
+	}
+
+	/**
+	 * Don't cache requests to scoped URLs or if the referrer URL is scoped.
+	 *
+	 * These requests are made to the PHP Worker Thread and are not static assets.
+	 */
+	if (isURLScoped(url)) {
+		return;
+	}
+	let referrerUrl;
+	try {
+		referrerUrl = new URL(event.request.referrer);
+	} catch (e) {
+		// ignore
+	}
+	if (referrerUrl && isURLScoped(referrerUrl)) {
+		return;
+	}
+
+	/**
+	 * Respond with cached assets if available.
+	 * If the asset is not cached, fetch it from the network and cache it.
+	 */
+	event.respondWith(cache.cachedFetch(event.request));
+});
```

Comment:
In the future, we could merge the fetch lister from `initializeServiceWorker` with this one.

I didn't want to do it now because I'm not sure if removing `initializeServiceWorker` will break backwards compatibility.